File: java-workshop/part-1-powermock/src/main/java/service/OrderCancellationService.java
package service;

import domain.CancellationResult;
import domain.Order;
import domain.OrderStatus;
import domain.StockReservation;
import repository.OrderRepository;

import java.util.List;

public class OrderCancellationService {

    public CancellationResult cancelOrder(String orderId) {
        PaymentService paymentService = new PaymentService();
        StockService stockService = new StockService();
        ShippingService shippingService = new ShippingService();

        Order order = OrderRepository.findById(orderId);
        if (order == null) {
            return new CancellationResult.Failure("Order not found");
        }

        if (order.transactionId() != null && !paymentService.refundPayment(order.transactionId())) {
            return new CancellationResult.Failure("Refund failed");
        }

        List<StockReservation> reservations = order.items().stream()
            .map(item -> new StockReservation(item.productId(), item.quantity(), true))
            .toList();
        stockService.releaseStock(reservations);

        if (order.trackingNumber() != null) {
            shippingService.cancelShipment(order.trackingNumber());
        }

        OrderRepository.updateStatus(orderId, OrderStatus.CANCELLED);
        return new CancellationResult.Success();
    }
}

File: csharp-workshop/Part1.StaticDeps/Service/OrderCancellationService.cs
using Part1.StaticDeps.Domain;
using Part1.StaticDeps.Repository;

namespace Part1.StaticDeps.Service;

public class OrderCancellationService
{
    public CancellationResult CancelOrder(string orderId)
    {
        var paymentService = new PaymentService();
        var stockService = new StockService();
        var shippingService = new ShippingService();
        var orderRepository = new OrderRepository();

        var order = orderRepository.FindById(orderId);
        if (order is null)
        {
            return new CancellationResult.Failure("Order not found");
        }

        if (order.TransactionId is not null && !paymentService.RefundPayment(order.TransactionId))
        {
            return new CancellationResult.Failure("Refund failed");
        }

        var reservations = order.Items
            .Select(item => new StockReservation(item.ProductId, item.Quantity, true))
            .ToList();
        stockService.ReleaseStock(reservations);

        if (order.TrackingNumber is not null)
        {
            shippingService.CancelShipment(order.TrackingNumber);
        }

        orderRepository.UpdateStatus(orderId, OrderStatus.Cancelled);
        return new CancellationResult.Success();
    }
}
